README.md
# EntriesTree.js

This library is intended to work with recursive array of data in Javascript.

- [Installation](#installation)
- [Guide](#guide)
- [API](#api)
- [Tests](#tests)
- [Licence](#licence)

## Installation

⬆️ [Top](#entriestreejs) ➡️ [Guide](#guide) 

This package is available on [NPM](https://www.npmjs.com/package/entriestree).

`npm i entriestree`

## Guide

⬆️ [Top](#entriestreejs) ➡️ [API](#api)

### Example data

This kind of data looks familiar? This is the sample data used below. You'll find this [sample data](https://github.com/MarceauKa/entriestree.js/blob/master/fixtures/readmeCollection.js) in `fixtures/readmeCollection.js`.

```
- {id: 1, ...}
	- {id: 10, ...}
	- {id: 11, ...}
		- {id: 110, ...}
		- {id: 111, ...}
			- {id: 1110, ...}
			- {id: 1111, ...}
		- {id: 112, ...}
	- {id: 12, ...}
- {id: 2, ...}
	- {id: 20, ...}
		- {id: 200, ...}
	- {id: 21, ...}
	- {id: 22, ...}
		- {id: 221, ...}
- {id: 3, ...}
- {id: 4, ...}
	- {id: 40, ...}
		- {id: 400, ...}
		- {id: 401, ...}
			- {id: 4010, ...}
	- {id: 41, ...}
	- {id: 42, ...}
```

### Initialization

In this example, `collection` is our structure and each item has an unique identifier called `id` (_customizable_). Child items are stored in a property called `children` (_customizable too_).

```js
import collection from '...' // above sample data
import EntriesTree from 'entriestree'

const tree = new EntriesTree(collection, 'id', 'children')
```

### Find element

Properties starting with `_` are added by the library.

```js
tree.find(11) // returns {id: 11, children: [{id: 110, ...}, ...]}
tree.find(3) // returns {id: 3, ...}
tree.find(-1) // returns null
```

Also, you can find the nearest ancestor.

```js
tree.findAncestor(110) // returns {id: 11, ...}
tree.findAncestor(1) // returns {id: 1, ...}
tree.findAncestor(-1) // returns null
```

### Find element's parent

```js
tree.findParent(110) // returns {id: 1, children: [...], ...}
tree.findParent(4) // returns null because {id: 4} is a root element
```

### Find element's siblings

```js
tree.findSiblings(1)
// returns {prevItem: null, nextItem: {id: 2, ...}}
tree.findSiblings(111)
// returns {prevItem: {id: 110, ...}, nextItem: {id: 112, ...}}
tree.findSiblings(-1)
// returns {prevItem: null, nextItem: null}
```

### Count elements

It's super easy to get the collection length (recursively).

```js
tree.count() // returns 23
tree.countFrom(11) // returns 5
````

### Update element

```js
tree.find(221) // returns {id: 221, ...}

tree.update(221, (item) => {
  item.foo = 'bar';
  return item
})

tree.find(221) // returns {id: 221, foo: 'bar', ...}
```

### Update element recursively

```js
tree.updateUp(110, (item) => {
  item.foo = 'bar'
  return item
}) // updates 110, 11 and 1 

tree.updateDown(111, (item) => {
  item.bar = 'baz'
  return item
}) // updates 111, 1110 and 1111 
```

⚠️ Don't forget to return the updated element in your updater function!

### Delete element

```js
tree.delete(111) // returns the deleted item
tree.count() // returns 20 since {id: 111} as two children
tree.delete(-1) // returns null
```

### Insert before or after element

```js
tree.insertAfter(1, toInsert) // returns EntriesTree's instance
tree.insertBefore(111, toInsert) // same as above
tree.insertAfter(-1, toInsert) // can't insert after an unexisting element
```

### Loop over flattened elements

```js
tree.iterable()
// returns [{id: 1, ...}, {id: 2, ...}, {id: 3, ...}, ...]
```

### Deep clone a collection

Sometimes it's useful to work on a copy of your elements without noising your original collection.

```js
const tree = new EntriesTree().setCollection(collection, true)
// collection will be untouched
const tree = new EntriesTree().clone(collection)
// same as above

tree.update(1, (item) => {
  item.foo = 'bar'
  return item
})

collection[0].foo !== tree.find(1).foo
// returns false
```

# API

⬆️ [Top](#entriestreejs) ➡️ [Tests](#tests)

Methods meant to be private are not listed. Eg: `isTheOne`, `isNode`, `isRoot`, etc

| Method | Params | Return |
|--------|--------|--------|
| `constructor` | `{Object[]} collection = []`, `{string} id = 'id'`, `{string} childkey` | `{EntriesTree}` |
| `setCollection` | `{Object[]} collection = []`, `{boolean} clone` | `{EntriesTree}` |
| `getCollection` |  | `{Object[]}` |
| `clone` | `{?Object[]} collection = null` | `{Object[]}` |
| `iterable` | `{Object[]}` |  |
| `count` | | `@returns {number}` |
| `countFrom` | `{Object,string,number} toFind` | `{number}` |
| `find` | `{Object,string,number} toFind` | `{?Object}` |
| `findParent` | `{Object,string,number} toFind` | `{?Object}` |
| `findSiblings` | `{Object,string,number} toFind` | `{?Object}` |
| `findAncestor` | `{Object,string,number} toFind` | `{?Object}` |
| `update` | `{Object,string,number} toFind`, `{function} updater` | `{EntriesTree}` |
| `updateUp` | `{Object,string,number} toFind`, `{function} updater` | `{EntriesTree}` |
| `updateDown` | `{Object,string,number} toFind`, `{function} updater` | `{EntriesTree}` |
| `delete` | `{Object,string,number} toFind` | `{?Object}` |
| `insertAfter` | `{Object,string,number} toFind`, `{Object} toInsert` | `{EntriesTree}` |
| `insertBefore` | `{Object,string,number} toFind`, `{Object} toInsert` | `{EntriesTree}` |

# Tests

⬆️ [Top](#entriestreejs) ➡️ [Licence](#licence)

Tests are made with [ava](https://github.com/avajs/ava)

`npm run test`

# Licence

⬆️ [Top](#entriestreejs)

MIT


CHANGELOG.md
# CHANGELOG

## 0.4.0 (2021-03-02)

- ⚠️ `siblings` method is now `findSiblings`
- ⚠️ `parent` method is now `findParent`
- Add `updateUp` and `updateDown` methods
- Performance improvements on `find` and `update` methods

## 0.3.0 (2021-02-27)

- Add `clone` method and `setCollection` now accepts a `clone` argument
- Add `countFrom` method
- Add `findAncestor` method
- Add `insertAfter` and `insertBefore` methods
- API description in readme
- Sample data used in readme can be found in `fixtures/readmeCollection.js`

## 0.2.2 (2021-02-23)

- Remove usage of ES properties

## 0.2.1 (2021-02-23)

- Remove usage of ES private properties
- Improved `.gitignore`

## 0.2.0 (2021-02-23)

- Add `iterable` method

## 0.1.0 (2021-02-19)

- Initial release with `count`, `find`, `update`, `delete`, `parent` and `siblings`
- Unit tests with the ava framework
- Published on NPM


tests/tests.js
import test from 'ava'
import defaultCollection from '../fixtures/defaultCollection.js'
import collectionWithCustomId from '../fixtures/collectionWithCustomId.js'
import collectionWithCustomChildren from '../fixtures/collectionWithCustomChildren.js'
import EntriesTree from '../src/EntriesTree.js'

test.beforeEach(t => {
  t.context.tree = new EntriesTree().setCollection(defaultCollection, true)
})

test('it walks through items', t => {
  let loop = 0

  for (let item of t.context.tree.iterable()) {
    t.is(item.id, loop + 1)
    loop++
  }

  t.is(t.context.tree.iterable().length, 14)
})

test('it finds root item', t => {
  t.is(t.context.tree.find(1).id, defaultCollection[0].id)
  t.is(t.context.tree.find(8).id, defaultCollection[1].children[1].children[0].id)
  t.is(t.context.tree.find(99), null)
  t.is(t.context.tree.find(defaultCollection[0].children[1]), t.context.tree.find(3))
})

test('it finds item with custom structure id', t => {
  const tree = new EntriesTree(collectionWithCustomId, 'uuid')
  t.is(tree.find('2210'), collectionWithCustomId[1].children[1].children[0])
})

test('it finds item with custom structure children', t => {
  const tree = new EntriesTree(collectionWithCustomChildren, 'id', 'blocks')
  t.is(tree.find(4), collectionWithCustomChildren[0].blocks[2])
})

test('it finds ancestor', t => {
  t.is(t.context.tree.find(1), t.context.tree.findAncestor(3))
  t.is(t.context.tree.find(1), t.context.tree.findAncestor(1))
  t.is(t.context.tree.find(5), t.context.tree.findAncestor(10))
  t.is(t.context.tree.find(99), null)
})

test('it count items', t => {
  t.is(t.context.tree.count(), 14)
})

test('it count items from item', t => {
  t.is(t.context.tree.countFrom(1), 3)
  t.is(t.context.tree.countFrom(99), 0)
  t.is(t.context.tree.countFrom(4), 0)
  t.is(t.context.tree.countFrom(9), 2)
})

test('it parentizes items', t => {
  t.is(t.context.tree.find(1)._parent, null)
  t.is(t.context.tree.find(1)._deepness, 0)

  t.is(t.context.tree.find(4)._parent, 1)
  t.is(t.context.tree.find(4)._deepness, 1)

  t.is(t.context.tree.find(10)._parent, 9)
  t.is(t.context.tree.find(10)._deepness, 3)
})

test('it clones tree', t => {
  let collection = [{ id: 1, children: [{ id: 2 }, { id: 3, children: [{ id: 4 }] }] }]
  let originalTree = new EntriesTree(collection)
  let copyTree = new EntriesTree(originalTree.clone())

  originalTree.update(2, (item) => {
    item.foo = 'bar'
    return item
  })

  t.is(originalTree.count(), copyTree.count())

  t.not(originalTree.find(2), copyTree.find(2))
  t.is(originalTree.find(2).id, copyTree.find(2).id)

  copyTree.update(4, (item) => {
    item.test = 'foo'
    return item
  })

  originalTree.update(4, (item) => {
    item.test = 'bar'
    return item
  })

  t.not(originalTree.find(4).test, copyTree.find(4).test)

  originalTree.delete(3)
  t.not(originalTree.count(), copyTree.count())
})

test('it finds parent item', t => {
  t.is(t.context.tree.findParent(1), null)
  t.is(t.context.tree.findParent(2), t.context.tree.find(1))
})

test('it finds siblings of an item', t => {
  let siblings = t.context.tree.findSiblings(1)
  t.is(siblings.prevItem, null)
  t.is(siblings.nextItem.id, 5)

  siblings = t.context.tree.findSiblings(5)
  t.is(siblings.prevItem.id, 1)
  t.is(siblings.nextItem.id, 12)

  siblings = t.context.tree.findSiblings(11)
  t.is(siblings.prevItem.id, 10)
  t.is(siblings.nextItem, null)
})

test('it finds siblings of an item from an object', t => {
  let siblings = t.context.tree.findSiblings(t.context.tree.find(5))
  t.is(siblings.prevItem.id, 1)
  t.is(siblings.nextItem.id, 12)

  siblings = t.context.tree.findSiblings(defaultCollection[1].children[1].children[1].children[0])
  t.is(siblings.prevItem, null)
  t.is(siblings.nextItem.id, 11)
})

test('it updates item', t => {
  t.context.tree.update(2, (item) => {
    item.type = 'updated'
    return item
  })
  t.is(t.context.tree.find(2).type, 'updated')
})

test('it updates item up', t => {
  t.context.tree.updateUp(8, (item) => {
    item.foo = 'bar'
    return item
  })

  t.is(t.context.tree.find(8).foo, 'bar')
  t.is(t.context.tree.find(7).foo, 'bar')
  t.is(t.context.tree.find(5).foo, 'bar')

  t.falsy(t.context.tree.find(1).hasOwnProperty('foo'))
  t.falsy(t.context.tree.find(6).hasOwnProperty('foo'))
})

test('it updates item down', t => {
  t.context.tree.updateDown(7, (item) => {
    item.bar = 'foo'
    return item
  })

  t.is(t.context.tree.find(7).bar, 'foo')
  t.is(t.context.tree.find(8).bar, 'foo')
  t.is(t.context.tree.find(10).bar, 'foo')

  t.falsy(t.context.tree.find(1).hasOwnProperty('bar'))
  t.falsy(t.context.tree.find(6).hasOwnProperty('bar'))
})

test('it deletes item', t => {
  t.not(t.context.tree.delete(12), null)
  t.is(t.context.tree.count(), 11)

  t.is(t.context.tree.delete(99), null)
  t.is(t.context.tree.count(), 11)

  t.not(t.context.tree.delete(4), null)
  t.is(t.context.tree.count(), 10)
})

test('it inserts item after child', t => {
  t.context.tree.insertAfter(2, { id: 222, message: 'added after first' })

  t.is(t.context.tree.countFrom(1), 4)
  t.is(t.context.tree.findSiblings(222).prevItem.id, 2)
  t.is(t.context.tree.findSiblings(222).nextItem.id, 3)
  t.truthy(t.context.tree.find(222).hasOwnProperty('_parent'))

  t.context.tree.insertAfter(4, { id: 444, message: 'added after last' })

  t.is(t.context.tree.countFrom(1), 5)
  t.is(t.context.tree.findSiblings(444).prevItem.id, 4)
  t.is(t.context.tree.findSiblings(444).nextItem, null)
  t.truthy(t.context.tree.find(444).hasOwnProperty('_parent'))
})

test('it inserts item after root', t => {
  t.context.tree.insertAfter(1, { id: 111, message: 'added after first' })

  t.is(t.context.tree.count(), 15)
  t.is(t.context.tree.findSiblings(111).prevItem.id, 1)
  t.is(t.context.tree.findSiblings(111).nextItem.id, 5)
  t.truthy(t.context.tree.find(11).hasOwnProperty('_parent'))

  t.context.tree.insertAfter(12, { id: 1212, message: 'added after last' })

  t.is(t.context.tree.count(), 16)
  t.is(t.context.tree.findSiblings(1212).prevItem.id, 12)
  t.is(t.context.tree.findSiblings(1212).nextItem, null)
  t.truthy(t.context.tree.find(1212).hasOwnProperty('_parent'))
})

test('it inserts item before child', t => {
  t.context.tree.insertBefore(2, { id: 222, type: 'added before first' })

  t.is(t.context.tree.count(), 15)
  t.is(t.context.tree.findSiblings(222).prevItem, null)
  t.is(t.context.tree.findSiblings(222).nextItem.id, 2)
  t.truthy(t.context.tree.find(222).hasOwnProperty('_parent'))

  t.context.tree.insertBefore(4, { id: 444, type: 'added before last' })

  t.is(t.context.tree.count(), 16)
  t.is(t.context.tree.findSiblings(444).prevItem.id, 3)
  t.is(t.context.tree.findSiblings(444).nextItem.id, 4)
  t.truthy(t.context.tree.find(444).hasOwnProperty('_parent'))
})

test('it inserts item before root', t => {
  t.context.tree.insertBefore(1, { id: 111, type: 'added before first' })

  t.is(t.context.tree.count(), 15)
  t.is(t.context.tree.findSiblings(111).prevItem, null)
  t.is(t.context.tree.findSiblings(111).nextItem.id, 1)
  t.truthy(t.context.tree.find(111).hasOwnProperty('_parent'))

  t.context.tree.insertBefore(12, { id: 1212, type: 'added before last' })

  t.is(t.context.tree.count(), 16)
  t.is(t.context.tree.findSiblings(1212).prevItem.id, 5)
  t.is(t.context.tree.findSiblings(1212).nextItem.id, 12)
  t.truthy(t.context.tree.find(1212).hasOwnProperty('_parent'))
})

demo.js
import demoCollection from './src/dummy.js'
import EntriesTree from './src/EntriesTree.js'

const tree = new EntriesTree(demoCollection)

// Find
console.group('Find')
console.log('find unexisting element should be null:', tree.find(-1))
console.log('find existing element:', tree.find(11))
console.groupEnd()

// Update
console.group('Update')
console.log('before update:', tree.find(3).type)
tree.update(3, (item) => {
	item.type = 'updated'
	return item
})
console.log('after update:', tree.find(3).type)
console.groupEnd()

// Count
console.group('Count')
console.log('count from method:', tree.count())
console.log('existing collection:', tree.collection)
console.groupEnd()

// Parent
console.group('Parent')
console.log('parent of root must be null:', tree.parent(1))
console.log('parent of children:', tree.parent(2))
console.groupEnd()

// Delete
console.group('Delete')
console.log('count before delete:', tree.count())
tree.delete(2)
console.log('count after delete:', tree.count())
console.groupEnd()


src/EntriesTree.js
export default class EntriesTree {
  /**
   * @constructor
   * @param {Object[]} collection
   * @param {string} itemKey
   * @param {string} childKey
   */
  constructor (collection = [], itemKey = 'id', childKey = 'children') {
    this.itemKey = itemKey
    this.childKey = childKey
    this.setCollection(collection)
  }

  /**
   * @returns {Object[]}
   */
  getCollection () {
    return this.collection
  }

  /**
   * @param {Object[]} collection
   * @param {boolean} clone
   * @returns {EntriesTree}
   */
  setCollection (collection, clone = false) {
    if (clone) {
      collection = this.clone(collection)
    }

    this.collection = this.parentize(collection)

    return this
  }

  /**
   * @param {?Object[]} collection
   * @return {Object[]}
   */
  clone (collection = null) {
    const cloner = (elements, stack = []) => {
      elements.forEach(item => {
        let copy = {}

        Object.keys(item).forEach(key => {
          let value = item[key]

          if (Object.prototype.toString.call(value) === '[object Object]' || Array.isArray(value)) {
            copy[key] = cloner(value)
          } else {
            copy[key] = value
          }
        })

        stack.push(copy)
      })

      return stack
    }

    return cloner(collection || this.collection, [])
  }

  /**
   * @param {Object|number|string} findable
   * @param {Object} comparable
   * @returns {boolean}
   */
  isTheOne (findable, comparable) {
    if (typeof findable === 'object') {
      return findable[this.itemKey] === comparable[this.itemKey]
    }

    return findable === comparable[this.itemKey]
  }

  /**
   * @param {Object} item
   * @returns {boolean}
   */
  isNode (item) {
    return item.hasOwnProperty(this.childKey)
  }

  /**
   * @param {Object} item
   * @returns {boolean}
   */
  isRoot (item) {
    return item.hasOwnProperty('_deepness') && item._deepness === 0
  }

  /**
   * @param {Object[]} elements
   * @returns {Object[]}
   */
  parentize (elements) {
    const invokable = (elements, parent = null) => {
      let stack = []

      elements.forEach(item => {
        item._parent = parent ? parent[this.itemKey] : null
        item._deepness = parent ? parent._deepness + 1 : 0

        if (this.isNode(item)) {
          item[this.childKey] = invokable(item[this.childKey], item)
        }

        stack.push(item)
      })

      return stack
    }

    return invokable(elements)
  }

  /**
   * @returns {Object[]}
   */
  iterable () {
    const invokable = (elements, parent = []) => {
      let stack = parent

      elements.forEach(item => {
        stack.push(item)

        if (this.isNode(item)) {
          stack = invokable(item[this.childKey], stack)
        }
      })

      return stack
    }

    return invokable(this.collection)
  }

  /**
   * @returns {number}
   */
  count () {
    const invokable = elements => {
      let total = 0

      elements.forEach(item => {
        total += 1

        if (this.isNode(item)) {
          total += invokable(item[this.childKey])
        }
      })

      return total
    }

    return invokable(this.collection)
  }

  /**
   * @param {Object|string|number} toFind
   * @returns {number}
   */
  countFrom (toFind) {
    toFind = this.find(toFind)

    const counter = (elements, total = 0) => {
      elements.forEach(item => {
        total += 1

        if (this.isNode(item)) {
          total += counter(item[this.childKey], total)
        }
      })

      return total
    }

    if (toFind && this.isNode(toFind)) {
      return counter(toFind[this.childKey])
    }

    return 0
  }

  /**
   * @param {Object|string|number} toFind
   * @returns {?Object}
   */
  find (toFind) {
    const invokable = (elements, parent = null) => {
      let found = null

      elements.some(item => {
        if (found !== null) {
          return false
        }

        if (this.isTheOne(toFind, item)) {
          found = item

          return false
        } else if (this.isNode(item)) {
          found = invokable(item[this.childKey], item)

          return found !== null
        }
      })

      return found
    }

    return invokable(this.collection)
  }

  /**
   * @param {Object|string|number} toFind
   * @returns {?Object}
   */
  findSiblings (toFind) {
    toFind = this.find(toFind)

    let siblings = {
      prevItem: null,
      nextItem: null
    }

    if (!toFind) {
      return siblings
    }

    let collection = toFind._parent ? this.find(toFind._parent)[this.childKey] : this.collection
    let index = collection.indexOf(toFind)

    if (index === -1) {
      return siblings
    }

    siblings.nextItem = collection[index + 1] || null
    siblings.prevItem = collection[index - 1] || null

    return siblings
  }

  /**
   * @param {Object|string|number} toFind
   * @returns {?Object}
   */
  findParent (toFind) {
    toFind = this.find(toFind)

    if (toFind && toFind._parent) {
      return this.find(toFind._parent)
    }

    return null
  }

  /**
   * @param {Object|string|number} toFind
   * @returns {?Object}
   */
  findAncestor (toFind) {
    toFind = this.find(toFind)

    const finder = element => {
      if (element._parent) {
        return finder(this.find(element._parent))
      }

      return element
    }

    if (toFind) {
      return finder(toFind)
    }

    return null
  }

  /**
   * @param {Object|string|number} toFind
   * @param {function} updater
   * @returns {EntriesTree}
   */
  update (toFind, updater) {
    toFind = this.find(toFind)

    const invokable = (elements) => {
      let stack = []

      elements.forEach(item => {
        if (this.isTheOne(toFind, item)) {
          item = updater(item)
        }

        if (this.isNode(item)) {
          item[this.childKey] = invokable(item[this.childKey])
        }

        stack.push(item)
      })

      return stack
    }

    if (toFind) {
      this.setCollection(invokable(this.collection))
    }

    return this
  }

  /**
   * @param {Object|string|number} toFind
   * @param {function} updater
   * @returns {EntriesTree}
   */
  updateUp (toFind, updater) {
    toFind = this.find(toFind)

    const invokable = (element, updater) => {
      element = updater(element)

      if (element._parent) {
        invokable(this.findParent(element), updater)
      }

      return element
    }

    if (toFind) {
      this.update(toFind, (item) => {
        invokable(item, updater)

        return item
      })
    }

    return this
  }

  /**
   * @param {Object|string|number} toFind
   * @param {function} updater
   * @returns {EntriesTree}
   */
  updateDown (toFind, updater) {
    toFind = this.find(toFind)

    const invokable = (element, updater) => {
      element = updater(element)

      if (this.isNode(element)) {
        element[this.childKey].forEach(item => {
          invokable(item, updater)
        })
      }

      return element
    }

    if (toFind) {
      this.update(toFind, (item) => {
        invokable(item, updater)

        return item
      })
    }

    return this
  }

  /**
   * @param {Object|string|number} toFind
   * @returns {?Object}
   */
  delete (toFind) {
    toFind = this.find(toFind)

    let deleted = null

    const invokable = (elements) => {
      let stack = []

      elements.forEach(item => {
        if (this.isTheOne(toFind, item)) {
          deleted = item

          return
        }

        if (this.isNode(item)) {
          item[this.childKey] = invokable(item[this.childKey])
        }

        stack.push(item)
      })

      return stack
    }

    if (toFind) {
      this.setCollection(invokable(this.collection))
    }

    return deleted
  }

  /**
   * @param {Object|string|number} toFind
   * @param {Object} toInsert
   * @returns {EntriesTree}
   */
  insertAfter (toFind, toInsert) {
    toFind = this.find(toFind)

    let parentCollection = toFind._parent ? this.findParent(toFind)[this.childKey] : this.collection
    parentCollection.splice(parentCollection.indexOf(toFind) + 1, 0, toInsert)

    if (toFind._parent) {
      this.update(toFind._parent, (item) => {
        item[this.childKey] = parentCollection
        return item
      })

      return this
    }

    return this.setCollection(parentCollection)
  }

  /**
   * @param {Object|string|number} toFind
   * @param {Object} toInsert
   * @returns {EntriesTree}
   */
  insertBefore (toFind, toInsert) {
    toFind = this.find(toFind)

    let parentCollection = toFind._parent ? this.findParent(toFind)[this.childKey] : this.collection
    parentCollection.splice(parentCollection.indexOf(toFind), 0, toInsert)

    if (toFind._parent) {
      this.update(toFind._parent, (item) => {
        item[this.childKey] = parentCollection
        return item
      })

      return this
    }

    return this.setCollection(parentCollection)
  }
}


fixtures/defaultCollection.js
const data = [
  {
    id: 1,
    type: 'container',
    children: [
      {
        id: 2,
        type: 'text'
      },
      {
        id: 3,
        type: 'image'
      },
      {
        id: 4,
        type: 'text'
      }
    ]
  },
  {
    id: 5,
    type: 'container',
    children: [
      {
        id: 6,
        type: 'image'
      },
      {
        id: 7,
        type: 'container',
        children: [
          {
            id: 8,
            type: 'text'
          },
          {
            id: 9,
            type: 'container',
            children: [
              {
                id: 10,
                type: 'image'
              },
              {
                id: 11,
                type: 'image'
              }
            ]
          }
        ],
      },
    ]
  },
  {
    id: 12,
    type: 'container',
    children: [
      {
        id: 13,
        type: 'text'
      },
      {
        id: 14,
        type: 'text'
      }
    ]
  }
]

export default data


fixtures/collectionWithCustomId.js
const data = [
  {
    uuid: '1000',
    type: 'container',
    children: [
      {
        uuid: '1100',
        type: 'text'
      },
      {
        uuid: '1200',
        type: 'image'
      },
      {
        uuid: '1300',
        type: 'text'
      }
    ]
  },
  {
    uuid: '2000',
    type: 'container',
    children: [
      {
        uuid: '2100',
        type: 'image'
      },
      {
        uuid: '2200',
        type: 'container',
        children: [
          {
            uuid: '2210',
            type: 'text'
          },
          {
            uuid: '2220',
            type: 'container',
            children: [
              {
                uuid: '2221',
                type: 'image'
              },
              {
                uuid: '2222',
                type: 'image'
              }
            ]
          }
        ],
      },
    ]
  },
  {
    uuid: 12,
    type: 'container',
    children: [
      {
        uuid: 13,
        type: 'text'
      },
      {
        uuid: 14,
        type: 'text'
      }
    ]
  }
]

export default data


fixtures/collectionWithCustomChildren.js
const data = [
  {
    id: 1,
    type: 'container',
    blocks: [
      {
        id: 2,
        type: 'text'
      },
      {
        id: 3,
        type: 'image'
      },
      {
        id: 4,
        type: 'text'
      }
    ]
  },
  {
    id: 5,
    type: 'container',
    blocks: [
      {
        id: 6,
        type: 'image'
      },
      {
        id: 7,
        type: 'container',
        blocks: [
          {
            id: 8,
            type: 'text'
          },
          {
            id: 9,
            type: 'container',
            blocks: [
              {
                id: 10,
                type: 'image'
              },
              {
                id: 11,
                type: 'image'
              }
            ]
          }
        ],
      },
    ]
  },
  {
    id: 12,
    type: 'container',
    blocks: [
      {
        id: 13,
        type: 'text'
      },
      {
        id: 14,
        type: 'text'
      }
    ]
  }
]

export default data


fixtures/readmeCollection.js
// Sample data used in README
const data = [
  {
    id: 1,
    children: [
      {
        id: 10,
      },
      {
        id: 11,
        children: [
          {
            id: 110,
          },
          {
            id: 111,
            children: [
              {
                id: 1110,
              },
              {
                id: 1111,
              },
            ]
          },
          {
            id: 112,
          },
        ]
      },
      {
        id: 12,
      },
    ]
  },
  {
    id: 2,
    children: [
      {
        id: 20,
        children: [
          {
            id: 200,
          },
        ]
      },
      {
        id: 21,
      },
      {
        id: 22,
        children: [
          {
            id: 221,
          },
        ]
      },
    ]
  },
  {
    id: 3,
  },
  {
    id: 4,
    children: [
      {
        id: 40,
        children: [
          {
            id: 400,
          },
          {
            id: 401,
            children: [
              {
                id: 4010,
              },
            ]
          },
        ]
      },
      {
        id: 41,
      },
      {
        id: 42,
      },
    ]
  },
]

export default data
